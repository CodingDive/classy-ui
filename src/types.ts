export interface IClass {
  id: string;
  classname: string;
  variant: string | null;
  shortName: string;
  css: (name: string) => string;
  variable: {
    value: string;
    originalValue: string;
  } | null;
}

export type IClasses = Record<string, IClass>;

export type IVariants = Record<string, string>;

export type IClassnames<T extends string> = Record<
  string,
  {
    variants?:
      | { [name: string]: string }
      | ((
          variables: IVariables<T>,
          utils: {
            negative: (value: { [key: string]: string }) => { [key: string]: string };
          },
        ) => IVariants);
    css: (name: string, value: string) => string;
  }
>;

export interface IEvaluatedClassnames {
  [name: string]: {
    variants: { [name: string]: string };
    variantsWithoutVariables: { [name: string]: string };
    css: (name: string, value: string) => string;
  };
}

export type IVariables<T extends string> = {
  [key in T]: { [variant: string]: string };
};

export interface IBaseConfig<T extends string, U = IVariables<T>> {
  variables: IVariables<T>;
  screens: {
    [name: string]: (css: string, variables: IVariables<T>) => string;
  };
  classnames: IClassnames<T>;
  themes?: {
    [name: string]: {
      [key in keyof U]: U[key];
    };
  };
}

export interface IConfig<T extends string, U = IVariables<T>> {
  variables?: {
    [key in T]: ((variables: IVariables<T>) => { [variant: string]: string }) | { [variant: string]: string };
  };
  screens?: {
    [name: string]: (css: string, variables: IVariables<T>) => string;
  };
  classnames?: IClassnames<T>;
  themes?: {
    [name: string]: {
      [key in keyof U]: U[key];
    };
  };
}

export interface IEvaluatedThemes {
  [variable: string]: {
    [key: string]: {
      [theme: string]: string;
    };
  };
}

export interface IEvaluatedConfig {
  variables: IVariables<any>;
  screens: {
    [name: string]: (css: string, variables: IVariables<any>) => string;
  };
  classnames: IEvaluatedClassnames;
  themeNames: string[];
  themes?: IEvaluatedThemes;
}

export interface IClassesByType {
  screens: {
    [type: string]: string[];
  };
  common: {
    [id: string]: string;
  };
  themeVariables: {
    [theme: string]: {
      [variable: string]: string;
    };
  };
  rootVariables: {
    [variable: string]: string;
  };
}

export interface IExtractedClass {
  id?: string;
  uid?: string;
  name: string;
  decorators: string[];
}

export interface IExtractedClasses {
  [uid: string]: IExtractedClass;
}
