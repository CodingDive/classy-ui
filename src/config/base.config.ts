import { IConfig } from '../types';

export const config: IConfig = {
  space: ['0', '0.5rem', '1rem', '2rem'],
  sizes: {
    '0': '0',
    '1': '0.25rem',
    '2': '0.5rem',
    '3': '0.75rem',
    '4': '1rem',
    '5': '1.25rem',
    '6': '1.5rem',
    '8': '2rem',
    '10': '2.5rem',
    '12': '3rem',
    '16': '4rem',
    '20': '5rem',
    '24': '6rem',
    '32': '8rem',
    '40': '10rem',
    '48': '12rem',
    '56': '14rem',
    '64': '16rem',
    auto: 'auto',
    px: '1px',
    '1/2': '50%',
    '1/3': '33.333333%',
    '2/3': '66.666667%',
    '1/4': '25%',
    '2/4': '50%',
    '3/4': '75%',
    '1/5': '20%',
    '2/5': '40%',
    '3/5': '60%',
    '4/5': '80%',
    '1/6': '16.666667%',
    '2/6': '33.333333%',
    '3/6': '50%',
    '4/6': '66.666667%',
    '5/6': '83.333333%',
    '1/12': '8.333333%',
    '2/12': '16.66667%',
    '3/12': '25%',
    '4/12': '33.33333%',
    '5/12': '41.66667%',
    '6/12': '50%',
    '7/12': '58.33333%',
    '8/12': '66.66667%',
    '9/12': '75%',
    '10/12': '83.33333%',
    '11/12': '91.66667%',
    full: '100%',
    'screen-width': '100vw',
    'screen-height': '100vh',
  },
  breakpoints: {
    sm: '640px',
    md: '768px',
    lg: '1024px',
    xl: '1280px',
  },
  colors: {
    transparent: 'transparent',
    black: '#000',
    white: '#FFF',
  },
  borderStyles: {
    solid: 'solid',
    dashed: 'dashed',
    dotted: 'dotted',
    double: 'double',
    none: 'none',
  },
  borderWidths: ['0', '1px', '2px', '4px'],
  borderColors: ({ colors }) => colors,
  radii: ['0', '.125rem', '.25rem', '.5rem', '50%'],
  fontSizes: ['0.75rem', '0.875rem', '1rem', '1.125rem', '1.25rem', '1.5rem', '1.875rem', '2.25rem', '3rem', '4rem'],
  fontWeights: ['100', '200', '300', '400', '500', '600', '600', '700', '800', '900'],
  fonts: {
    sans:
      '-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"',
    serif: 'Georgia, Cambria, "Times New Roman", Times, serif',
    mono: 'Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace',
  },
  letterSpacings: ['-0.05em', '-0.025em', '0', '0.025em', '0.05em', '0.1em'],
  lineHeights: ['1', '1.25', '1.375', '1.5', '1.625', '2'],
  shadows: {
    default: '0 1px 3px 0 rgba(0, 0, 0, 0.1), 0 1px 2px 0 rgba(0, 0, 0, 0.06)',
    md: '0 4px 6px -1px rgba(0, 0, 0, 0.1), 0 2px 4px -1px rgba(0, 0, 0, 0.06)',
    lg: '0 10px 15px -3px rgba(0, 0, 0, 0.1), 0 4px 6px -2px rgba(0, 0, 0, 0.05)',
    xl: '0 20px 25px -5px rgba(0, 0, 0, 0.1), 0 10px 10px -5px rgba(0, 0, 0, 0.04)',
    inner: 'inset 0 2px 4px 0 rgba(0, 0, 0, 0.06)',
    outline: '0 0 0 3px rgba(66, 153, 225, 0.5)',
    none: 'none',
  },
  transitions: {
    'fade-in': 'opacity 0.5s ease-in',
  },
  zIndices: [0, 10, 20, 30, 40],
};
