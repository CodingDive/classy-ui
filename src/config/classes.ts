import { IConfig, TClasses } from '../types';

export const classes = (
  config: IConfig,
): {
  [key in TClasses]?: string;
} => ({
  'bg-transparent': `{background-color: ${config.backgrounds.backgroundColor.transparent}`,
  'bg-black': `{background-color: ${config.backgrounds.backgroundColor.black}`,
  'bg-white': `{background-color: ${config.backgrounds.backgroundColor.white}`,
  'bg-gray-100': `{background-color: ${config.backgrounds.backgroundColor['gray-100']}`,
  'bg-gray-200': `{background-color: ${config.backgrounds.backgroundColor['gray-200']}`,
  'bg-gray-300': `{background-color: ${config.backgrounds.backgroundColor['gray-300']}`,
  'bg-gray-400': `{background-color: ${config.backgrounds.backgroundColor['gray-400']}`,
  'bg-gray-500': `{background-color: ${config.backgrounds.backgroundColor['gray-500']}`,
  'bg-gray-600': `{background-color: ${config.backgrounds.backgroundColor['gray-600']}`,
  'bg-gray-700': `{background-color: ${config.backgrounds.backgroundColor['gray-700']}`,
  'bg-gray-800': `{background-color: ${config.backgrounds.backgroundColor['gray-800']}`,
  'bg-gray-900': `{background-color: ${config.backgrounds.backgroundColor['gray-900']}`,
});
