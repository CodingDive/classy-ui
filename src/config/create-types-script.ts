import { writeFileSync } from 'fs';
import { join } from 'path';

import { evaluateConfig } from '../utils';

import { transform as transformClassesToTypes } from './transform-classes-to-types';

const config = evaluateConfig({});

if (process.env.NODE_ENV !== 'test') {
  const libTypesPath = join(process.cwd(), 'lib', 'classy-ui.d.ts');
  const types = transformClassesToTypes(config);
  writeFileSync(libTypesPath, types);
}
