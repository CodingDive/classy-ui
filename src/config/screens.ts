import { IScreens } from '../types';

const screens: IScreens = {
  mobile: (css, breakpoints) => `@media (min-width: ${breakpoints.MOBILE}) {${css}}`,
  tablet: (css, breakpoints) => `@media (min-width: ${breakpoints.TABLET}) {${css}}`,
  laptop: (css, breakpoints) => `@media (min-width: ${breakpoints.LAPTOP}) {${css}}`,
  desktop: (css, breakpoints) => `@media (min-width: ${breakpoints.DESKTOP}) {${css}}`,
};

export default screens;
