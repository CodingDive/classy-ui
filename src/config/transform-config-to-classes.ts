import { IClasses, IConfig, IEvaluatedConfig, TCssClasses } from '../types';
import { generateShortName, getClassesFromConfig } from '../utils';

const cssClasses: TCssClasses = {
  padding: ['padding', 'padding-top', 'padding-right', 'padding-bottom', 'padding-left'],
  margin: ['margin', 'margin-top', 'margin-right', 'margin-bottom', 'margin-left'],
  placeholderColor: [],
  gap: ['gap'],
  rowGap: ['row-gap'],
  columnGap: ['column-gap'],
  gridTemplateColumns: ['grid-template-columns'],
  gridColumnStart: ['grid-column-start'],
  gridColumnEnd: ['grid-column-end'],
  gridTemplateRows: ['grid-template-rows'],
  gridRow: ['grid-row'],
  gridRowStart: ['grid-row-start'],
  gridRowEnd: ['grid-row-end'],
  transformOrigin: ['transform-origin'],
  scale: ['scale'],
  rotate: ['rotate'],
  translate: ['translate'],
  skew: ['skew'],
  transitionProperty: ['transition-property'],
  transitionTimingFunction: ['transition-timing-function'],
  transitionDuration: ['transition-duration'],
  gridColumn: ['grid-column'],
  width: ['width'],
  borderWidth: ['border-width', 'border-top-width', 'border-right-width', 'border-bottom-width', 'border-left-width'],
  borderColor: ['border-color', 'border-bottom-color', 'border-top-color', 'border-left-color', 'border-right-color'],
  borderRadius: [
    'border-radius',
    'border-top-left-radius',
    'border-top-right-radius',
    'border-bottom-right-radius',
    'border-bottom-left-radius',
  ],
  fontSize: ['font-size'],
  fontWeight: ['font-weight'],
  fontFamily: ['font-family'],
  letterSpacing: ['letter-spacing'],
  lineHeight: ['line-height'],
  boxShadow: ['box-shadow', 'text-shadow'],
  zIndex: ['z-index'],
  backgroundPosition: ['background-position'],
  backgroundSize: ['background-size'],
  cursor: ['cursor'],
  fill: ['fill'],
  flex: ['flex'],
  inset: ['inset'],
  listStyleType: ['list-style-type'],
  objectPosition: ['object-position'],
  opacity: ['opacity'],
  order: ['order'],
  stroke: ['stroke'],
  strokeWidth: ['stroke-width'],
  backgroundColor: ['background-color'],
  container: [],
  flexGrow: ['flex-grow'],
  flexShrink: ['flex-shrink'],
  height: ['height'],
  maxHeight: ['max-height'],
  maxWidth: ['max-width'],
  minHeight: ['min-height'],
  minWidth: ['min-width'],
  color: ['color'],
};

export const transform = (config: IEvaluatedConfig): { defaults: IClasses; themes: { [name: string]: boolean } } => {
  const keys = Object.keys(cssClasses) as Array<keyof TCssClasses>;
  let cssPropertyCount = 0;
  const defaults = keys.reduce((aggr, key) => {
    return {
      ...aggr,
      ...getClassesFromConfig(
        key,
        config,
        cssClasses[key],
        (labelIndex: number) => `${generateShortName(cssPropertyCount++)}-${generateShortName(labelIndex)}`,
      ),
    };
  }, {} as IClasses);

  return {
    defaults,
    themes: Object.keys(config.themes || {}).reduce((aggr, theme) => {
      aggr[`themes-${theme}`] = true;

      return aggr;
    }, {} as any),
  };
};
