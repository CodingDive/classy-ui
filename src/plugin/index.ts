import { config } from '../config/base.config';
import { transform as transformConfigToClasses } from '../config/transform-config-to-classes';
import { getUserConfig, isBreakpoint, mergeConfigs } from '../utils';

const classes = transformConfigToClasses(mergeConfigs(config, getUserConfig()));

export default (babel: any) => {
  return {
    name: 'classy-ui/plugin',
    visitor: {
      Program(programmPath: any, state: any) {
        programmPath.traverse({
          ImportDeclaration(path: any) {
            if (path?.node?.source?.value === 'classy-ui') {
              const binding = path.scope.getBinding('classnames');

              if (binding && Boolean(binding.referencePaths.length)) {
                processReferences(babel, state, binding.referencePaths);
              }
            }
          },
        });
      },
    },
  };
};

function camleToDash(string: string) {
  return string
    .replace(/[\w]([A-Z])/g, function(m) {
      return m[0] + '-' + m[1];
    })
    .toLowerCase();
}

function generateShortName(number: number) {
  let baseChar = 'A'.charCodeAt(0);
  let letters = '';

  do {
    number -= 1;
    letters = String.fromCharCode(baseChar + (number % 26)) + letters;
    number = (number / 26) >> 0;
  } while (number > 0);

  return letters;
}

let nameCounter = 1;
const nameCache = new Map();

function computeName(id: string, isProduction: boolean) {
  if (isProduction) {
    if (nameCache.has(id)) {
      return nameCache.get(id);
    } else {
      const name = generateShortName(nameCounter++);
      nameCache.set(id, name);
      return name;
    }
  } else {
    return id;
  }
}

export function processReferences(babel: any, state: any, classnamesRefs: any) {
  const { types: t } = babel;

  const isProduction = babel.getEnv() === 'production';

  function convertToExpression(arr: any[]) {
    if (arr.length == 1) {
      return arr[0];
    } else {
      const strings = [];
      const others = [];
      for (let item of arr) {
        if (t.isStringLiteral(item)) {
          strings.push(item.value);
        } else {
          others.push(item);
        }
      }
      if (strings.length > 0 && others.length > 0) {
        others.unshift(t.stringLiteral(strings.join(' ')));
      } else if (strings.length > 0) {
        return t.stringLiteral(strings.join(' '));
      }

      let max = others.length - 1;
      let start = others[max];
      for (let i = max - 1; i >= 0; i--) {
        start = t.binaryExpression('+', others[i], start);
      }

      return start;
    }
  }

  function getImportName(name: string, scope: any) {
    const binding = scope.getBinding(name);
    if (binding && t.isImportSpecifier(binding.path.node) && binding.path.parent.source.value.startsWith('classy-ui')) {
      return binding.path.node.imported.name;
    }
    return null;
  }

  function createClassObject(id: string, { pseudos, breakpoints }: { pseudos: string[]; breakpoints: string[] }) {
    const name = computeName(id, isProduction);
    return { id, name, pseudos: pseudos.slice(), breakpoints: breakpoints.slice() };
  }

  function updateContext({ pseudos, breakpoints }: { pseudos: string[]; breakpoints: string[] }, value: string) {
    const context = { pseudos: pseudos.slice(), breakpoints: breakpoints.slice() };
    value = camleToDash(value);
    if (isBreakpoint(value)) {
      context.breakpoints.push(value);
    } else {
      context.pseudos.push(value);
    }
    return context;
  }

  function throwCodeFragmentIfIdNotFound(path: any, id: string) {
    if (!classes[id]) {
      throw path.buildCodeFrameError(`Could not find class ${id}`);
    }
  }

  function getIdOrThrow(path: any) {
    if (t.isStringLiteral(path.node)) {
      return path.node.value;
    } else if (t.isIdentifier(path.node)) {
      return path.node.name;
    } else {
      throw path.buildCodeFrameError(`This property is not supported`);
    }
  }

  function rewriteAndCollectArguments(context: any, argumentPaths: any, collect: any) {
    return argumentPaths.reduce((aggr: any[], argPath: any) => {
      const node = argPath.node;

      if (t.isStringLiteral(node)) {
        const classObj = createClassObject(node.value, context);
        throwCodeFragmentIfIdNotFound(argPath, classObj.id);
        collect.add(classObj);
        return aggr.concat([t.stringLiteral(classObj.name)]);
      } else if (t.isIdentifier(node)) {
        return aggr.concat([node]);
      } else if (t.isCallExpression(node) && t.isIdentifier(node.callee)) {
        // To support import { hover as ho }...
        const name = getImportName(node.callee.name, argPath.scope);
        // When not found it is not part of classy ui
        if (name != null) {
          const newContext = updateContext(context, name);
          return aggr.concat(rewriteAndCollectArguments(newContext, argPath.get('arguments'), collect));
        }
      } else if (t.isObjectExpression(node)) {
        return argPath.get('properties').map((propPath: any) => {
          const id = getIdOrThrow(propPath.get('key'));
          const classObj = createClassObject(id, context);
          throwCodeFragmentIfIdNotFound(propPath, classObj.id);
          collect.add(classObj);
          return aggr.concat(
            t.conditionalExpression(propPath.node.value, t.stringLiteral(classObj.name), t.stringLiteral('')),
          );
        });
      }

      return aggr.concat(node);
    }, []);
  }

  const classCollection = new Set();
  classnamesRefs
    .map((ref: any) => ref.findParent((p: any) => t.isCallExpression(p)))
    .forEach((call: any) => {
      const rewrite = rewriteAndCollectArguments(
        { pseudos: [], breakpoints: [] },
        call.get('arguments'),
        classCollection,
      );
      const newExpression = convertToExpression(rewrite.flat());
      if (newExpression) {
        call.replaceWith(newExpression);
      } else {
        call.remove();
      }
    });

  // TODO: Implement generating instead of throwing.
  throw new Error(JSON.stringify([...classCollection], null, 2));
}
