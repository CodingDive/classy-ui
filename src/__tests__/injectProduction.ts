import { transform } from '../config/transform-config-to-classes';
import { productionClassesByType } from '../plugin';
import { IClassesByType, IExtractedClasses } from '../types';
import { createClassObject, createProductionCss, evaluateConfig, injectProduction } from '../utils';

const config = evaluateConfig(require('../../classy-ui.config.js'));
const classes = transform(config);

describe('INJECT PRODUCTION', () => {
  beforeEach(() => {
    productionClassesByType.common = {};
    productionClassesByType.rootTokens = {};
    productionClassesByType.screens = {};
    productionClassesByType.themeTokens = {};
  });
  test('should inject simple', () => {
    const classCollection: IExtractedClasses = {
      ['color-RED']: createClassObject({ baseClass: 'color', token: 'RED', decorators: [] }, classes, true),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject with pseudo selector', () => {
    const classCollection: IExtractedClasses = {
      ['color-RED']: createClassObject({ baseClass: 'color', token: 'RED', decorators: ['hover'] }, classes, true),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject with breakpoint', () => {
    const classCollection: IExtractedClasses = {
      ['color-RED']: createClassObject({ baseClass: 'color', token: 'RED', decorators: ['tablet'] }, classes, true),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject both pseudo selector and breakpoint', () => {
    const classCollection: IExtractedClasses = {
      ['color-RED']: createClassObject(
        { baseClass: 'color', token: 'RED', decorators: ['hover', 'tablet'] },
        classes,
        true,
      ),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject with multiple pseudo selectors', () => {
    const classCollection: IExtractedClasses = {
      ['color-RED']: createClassObject(
        { baseClass: 'color', token: 'RED', decorators: ['hover', 'firstChild'] },
        classes,
        true,
      ),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject with multiple breakpoints', () => {
    const classCollection: IExtractedClasses = {
      ['color-RED']: createClassObject(
        { baseClass: 'color', token: 'RED', decorators: ['mobile', 'laptop'] },
        classes,
        true,
      ),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject multiple', () => {
    const classCollection: IExtractedClasses = {
      ['color-RED']: createClassObject({ baseClass: 'color', token: 'RED', decorators: ['tablet'] }, classes, true),
      ['color-BLUE']: createClassObject(
        { baseClass: 'color', token: 'BLUE', decorators: ['hover', 'tablet'] },
        classes,
        true,
      ),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should inject themes', () => {
    const classCollection: IExtractedClasses = {
      ['color-GREEN']: createClassObject({ baseClass: 'color', token: 'GREEN', decorators: ['tablet'] }, classes, true),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should group breakpoints', () => {
    const classCollection: IExtractedClasses = {
      ['color-RED']: createClassObject({ baseClass: 'color', token: 'RED', decorators: ['tablet'] }, classes, true),
      ['color-BLUE']: createClassObject({ baseClass: 'color', token: 'BLUE', decorators: ['tablet'] }, classes, true),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should group themes', () => {
    const classCollection: IExtractedClasses = {
      ['color-GREEN']: createClassObject({ baseClass: 'color', token: 'GREEN', decorators: [] }, classes, true),
      ['color-PURPLE']: createClassObject({ baseClass: 'color', token: 'PURPLE', decorators: [] }, classes, true),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should group themes and breakpoints', () => {
    const classCollection: IExtractedClasses = {
      ['color-GREEN']: createClassObject({ baseClass: 'color', token: 'RED', decorators: ['tablet'] }, classes, true),
      ['color-PURPLE']: createClassObject(
        { baseClass: 'color', token: 'PURPLE', decorators: ['tablet'] },
        classes,
        true,
      ),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should handle references classnames', () => {
    const classCollection: IExtractedClasses = {
      ['color-GREEN']: createClassObject(
        { baseClass: 'borderColor', token: 'GREEN', decorators: ['hover'] },
        classes,
        false,
      ),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
  test('should not override with screens', () => {
    const classCollection: IExtractedClasses = {
      ['color-RED']: createClassObject({ baseClass: 'color', token: 'RED', decorators: [] }, classes, true),
      ['color-GREEN']: createClassObject({ baseClass: 'color', token: 'GREEN', decorators: ['laptop'] }, classes, true),
    };
    const productionClassesByType: IClassesByType = {
      screens: {},
      common: {},
      themeTokens: {},
      rootTokens: {},
    };

    expect(
      createProductionCss(injectProduction(productionClassesByType, classCollection, classes, config), config),
    ).toMatchSnapshot();
  });
});
