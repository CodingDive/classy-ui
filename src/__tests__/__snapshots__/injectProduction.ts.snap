// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`INJECT PRODUCTION should group breakpoints 1`] = `
"@media (max-width: 768px) {.AV\\\\@A {
  color: red;
}.AV\\\\@B {
  color: blue;
}}"
`;

exports[`INJECT PRODUCTION should group themes 1`] = `
".AV\\\\@C {
  color: var(--colors-GREEN);
}.AV\\\\@D {
  color: var(--colors-PURPLE);
}:root{--colors-GREEN:green;--colors-PURPLE:purple;}.themes-dark{--colors-GREEN:yellow;--colors-PURPLE:pink;}"
`;

exports[`INJECT PRODUCTION should group themes and breakpoints 1`] = `
"@media (max-width: 768px) {.AV\\\\@A {
  color: red;
}.AV\\\\@D {
  color: var(--colors-PURPLE);
}}:root{--colors-PURPLE:purple;}.themes-dark{--colors-PURPLE:pink;}"
`;

exports[`INJECT PRODUCTION should handle references classnames 1`] = `
".P\\\\@C:hover {
  border-top-color: var(--colors-GREEN);
}.Q\\\\@C:hover {
  border-right-color: var(--colors-GREEN);
}.R\\\\@C:hover {
  border-bottom-color: var(--colors-GREEN);
}.S\\\\@C:hover {
  border-left-color: var(--colors-GREEN);
}:root{--colors-GREEN:green;}.themes-dark{--colors-GREEN:yellow;}"
`;

exports[`INJECT PRODUCTION should inject both pseudo selector and breakpoint 1`] = `
"@media (max-width: 768px) {.AV\\\\@A:hover {
  color: red;
}}"
`;

exports[`INJECT PRODUCTION should inject multiple 1`] = `
"@media (max-width: 768px) {.AV\\\\@A {
  color: red;
}.AV\\\\@B:hover {
  color: blue;
}}"
`;

exports[`INJECT PRODUCTION should inject simple 1`] = `
".AV\\\\@A {
  color: red;
}"
`;

exports[`INJECT PRODUCTION should inject themes 1`] = `
"@media (max-width: 768px) {.AV\\\\@C {
  color: var(--colors-GREEN);
}}:root{--colors-GREEN:green;}.themes-dark{--colors-GREEN:yellow;}"
`;

exports[`INJECT PRODUCTION should inject with breakpoint 1`] = `
"@media (max-width: 768px) {.AV\\\\@A {
  color: red;
}}"
`;

exports[`INJECT PRODUCTION should inject with multiple breakpoints 1`] = `
"@media (max-width: 640px) {.AV\\\\@A {
  color: red;
}}@media (max-width: 1280px) {.AV\\\\@A {
  color: red;
}}"
`;

exports[`INJECT PRODUCTION should inject with multiple pseudo selectors 1`] = `
".AV\\\\@A:first-child:hover {
  color: red;
}"
`;

exports[`INJECT PRODUCTION should inject with pseudo selector 1`] = `
".AV\\\\@A:hover {
  color: red;
}"
`;
